cephfs: safeguard localClusterState struct from race conditions

Rakshith-R opened the pull request:
# Describe what this PR does #

This commit uses atomic.Int64 and sync.Map with members of localClusterState and safeguards clusterAdditionalInfo map operations with a mutex.


Is the change backward compatible?

yes

Are there concerns around backward compatibility?

no

## Related issues ##

Fixes: #4162

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

Rakshith-R commented:
/test ci/centos/mini-e2e/k8s-1.27

Madhu-1 commented:
This fixes https://github.com/ceph/ceph-csi/issues/4162. feel free to attach the issue to the PR

Madhu-1 commented on internal/cephfs/core/volume.go:
```diff
@@ -38,6 +40,7 @@ import (
 // Subvolumegroup creation and volume resize decisions are
 // taken through this additional cluster information.
 var clusterAdditionalInfo = make(map[string]*localClusterState)
+var clusterAdditionalInfoMutex sync.Mutex
```
Early suggestion:- use sync.Map instead of map and sync.Mutex

Rakshith-R replied:
It'll make things more complex.

We'll need to get variable and convert to clusterState type every time and then store it back.
The race condition happens only when we are writing so current code should be good.
The upper map uses a mutex and members themselves are atomic.

Rakshith-R commented:
/retest ci/centos/mini-e2e/k8s-1.27

Madhu-1 commented on internal/cephfs/core/volume.go:
```diff
@@ -190,33 +197,32 @@ func (s *subVolumeClient) GetSubVolumeInfo(ctx context.Context) (*Subvolume, err
-
 const (
-	unknown operationState = iota
+	unknown int64 = iota
```
not sure why this was int64 before, as we have minimal values we should have used uint32.

Madhu-1 commented on internal/cephfs/core/volume.go:
```diff
@@ -297,8 +303,8 @@ func (s *subVolumeClient) ResizeVolume(ctx context.Context, bytesQuota int64) er
 	// clusterID key was not present.
-	if clusterAdditionalInfo[s.clusterID].resizeState == unknown ||
-		clusterAdditionalInfo[s.clusterID].resizeState == supported {
+	if clusterAdditionalInfo[s.clusterID].resizeState.Load() == unknown ||
+		clusterAdditionalInfo[s.clusterID].resizeState.Load() == supported {
```
This should also go to the helper function like in metadata.go as an enhancement.

nixpanic commented on internal/cephfs/core/volume.go:
```diff
@@ -32,12 +34,17 @@ import (
+	clusterAdditionalInfo = make(map[string]*localClusterState)
+	// clusterAdditionalInfoMutex is used to synchronize access to
+	// clusterAdditionalInfo map.
+	clusterAdditionalInfoMutex = sync.Mutex{}
```
Would a `sync.RWMutex` not be better?

If you only mean to protect against concurrent writes, please state it clearly.

nixpanic commented on internal/cephfs/core/volume.go:
```diff
@@ -232,7 +238,7 @@ func (s *subVolumeClient) CreateVolume(ctx context.Context) error {
 
 	// create subvolumegroup if not already created for the cluster.
-	if !clusterAdditionalInfo[s.clusterID].subVolumeGroupsCreated[s.FsName] {
+	if _, found := clusterAdditionalInfo[s.clusterID].subVolumeGroupsCreated.Load(s.FsName); !found {
```
This is still racy. If the subVolumeGroup is not found, there is a timespan where multiple processes can try to create it. This is something that should be protected by a Mutex.

If creating the subVolumeGroup multiple times is not problematic, please leave a comment in the code about it.

nixpanic commented on internal/cephfs/core/volume.go:
```diff
@@ -190,33 +197,32 @@ func (s *subVolumeClient) GetSubVolumeInfo(ctx context.Context) (*Subvolume, err
-	resizeState                 operationState
-	subVolMetadataState         operationState
-	subVolSnapshotMetadataState operationState
+	resizeState                 atomic.Int64
```
I do not understand why using an atomic int is safer here. Can you explain that in the commit message?

Rakshith-R commented:
On deeper inspection, indeed atomic really didn't serve the purpose and subVolgroupCreated was still kinda buggy. 

Rewrote the code to use RWMutex locks individually with lot of helpers.

Please take a look again.

Rakshith-R commented:
/retest ci/centos/mini-e2e/k8s-1.27

nixpanic commented on internal/cephfs/core/volume.go:
```diff
@@ -201,19 +212,22 @@ const (
-	subVolSnapshotMetadataState operationState
+	resizeState                 operationStateMutex
+	subVolMetadataState         operationStateMutex
+	subVolSnapshotMetadataState operationStateMutex
```
this may be overdoing it, is the `subVolumeGroupsRWMutex` of this struct not sufficient to protect its members?

Rakshith-R replied:
> this may be overdoing it, is the `subVolumeGroupsRWMutex` of this struct not sufficient to protect its members?

No, all are independent parameters.
+ with R-Locks it should basically not have much overhead.

nixpanic replied:
It is not about the performance overhead. I am worried about the complexity that looks unneeded.

nixpanic commented on internal/cephfs/core/volume.go:
```diff
@@ -190,7 +196,12 @@ func (s *subVolumeClient) GetSubVolumeInfo(ctx context.Context) (*Subvolume, err
-type operationState int64
+type operationState int32
+
+type operationStateMutex struct {
```
This seems to be introduced to prevent concurrent access to the int32. Are there really issues with reading/writing the values concurrently?

Rakshith-R replied:
> This seems to be introduced to prevent concurrent access to the int32. Are there really issues with reading/writing the values concurrently?

None have been seen, but being precautionary here.

nixpanic replied:
I don't see a problem with normal int usage. If there is no explanation of why it could be problematic, don't add the extra (unneeded) complexity.

nixpanic commented on internal/cephfs/core/volume.go:
```diff
@@ -201,19 +212,22 @@ const (
 	// set true once a subvolumegroup is created
 	// for corresponding filesystem in a cluster.
 	subVolumeGroupsCreated map[string]bool
+	subVolumeGroupsRWMutex sync.RWMutex
```
describe the usage of this mutex, what does it protect, when do callers need to lock it?

nixpanic reviewed (commented):
Without description of the locking done, it is a little guessing what the actual flow is, and what race condition needs to be prevented.

Can you summarize the problem description in the commit message in a short paragraph, and the implemented solution in an other?

nixpanic reviewed (commented):
You are currently changing two different things here. It may give more clarity when you split them up in two commits.

1. concurrent access to `clusterAdditionalInfo`
2. `operationState` type and members

I can understand a problem with 1, but for 2 I do not see what the issue is with the current code.

Rakshith-R committed:
cephfs: safeguard localClusterState struct from race conditions

Multiple go-routines may simultaneously check for a clusterID's
presence in clusterAdditionalInfo and create an entry if it is
absent. This set of operation needs to be serialized.

Therefore, this commit safeguards clusterAdditionalInfo map
from concurrent writes with a mutex to prevent the above problem.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
I've separated the changes into two different commits with explanations of issue and solution as requested.
PTAL

Rakshith-R commented:
/retest ci/centos/mini-e2e/k8s-1.27

Rakshith-R commented:
@mergifyio rebase

nixpanic commented on internal/cephfs/core/volume.go:
```diff
@@ -240,7 +247,7 @@ func (s *subVolumeClient) CreateVolume(ctx context.Context) error {
 
 	// create subvolumegroup if not already created for the cluster.
-	if !clusterAdditionalInfo[s.clusterID].subVolumeGroupsCreated[s.FsName] {
+	if !s.isSubVolumeGroupCreated(s.SubvolumeGroup) {
```
This is still racy. two or more go routines can call the function, and both can continue to create the SubVolumeGroup.

In order to prevent races, you will need to take a lock on the `clusterAdditionalInfo[s.clusterID]` object (a single lock for all members to keep it simple would be fine), or

1. lock the `.subVolumeGroupsRWMutex`
2. create the SubVolumeGroup
3. unlock the `.subVolumeGroupsRWMutex`

If access to the different members needs similar serialization, use a single Mutex to update any of the members and have other go routines block until the updating is done.

Rakshith-R replied:
> This is still racy. two or more go routines can call the function, and both can continue to create the SubVolumeGroup.
> 
> In order to prevent races, you will need to take a lock on the `clusterAdditionalInfo[s.clusterID]` object (a single lock for all members to keep it simple would be fine), or
> 
> 1. lock the `.subVolumeGroupsRWMutex`
> 2. create the SubVolumeGroup
> 3. unlock the `.subVolumeGroupsRWMutex`
> 
> If access to the different members needs similar serialization, use a single Mutex to update any of the members and have other go routines block until the updating is done.

Yes that is intentional, 
The mentioned parallel create will happen only once and there's no side affects to svg create cmd being called more than once. The follow on update will still be guarded by rwlock so we're fine. 
I don't want to hold locks while we wait for a response from ceph.
And I want don't want other members to wait while one is being updated.

nixpanic replied:
Sorry, but that is not what the commit message describes:

```
Multiple go-routines may simultaneously check for
presence of a clusterID's metadata such as
subvolumegroup created state, resize state,
metadata state and snapshot metadata state in the
clusterAdditionalInfo map and update an entry after creation
if it is absent. This set of operation needs to be serialized.
```

There is no serialization in this commit. This commit add atomically setting/reading a value. Usually an int is atomic already, and sufficient if there are no calculations done with it. I do not think the mutex per member in the `operationState` contributes to preventing race conditions.

Rakshith-R committed:
cephfs: safeguard subVolumeGroupCreated map from race condition

Multiple go-routines may simultaneously create the
subVolumeGroupCreated map or  write into it
for a particular group.

This commit safeguards subVolumeGroupCreated map
from concurrent creation/writes while allowing for multiple
readers.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
:+1: 
Modified second commit to protect only subVolumeGroupCreated map from consurrent creation/writes while allowing multiple readers.
PTAL

riya-singhal31 commented:
@Mergifyio queue

Rakshith-R commented:
@mergifyio backport release-3.9

Rakshith-R commented:
@Mergifyio backport release-v3.9
